modules/webbrowser/include/modules/webbrowser/processors/webbrowserprocessor.h

```diff
@@ -94,75 +74,41 @@ namespace inviwo {
  * </script>
  * \endcode
  *
- * ### Inports
- *   * __background__ Background to render web page ontop of.
- *
- * ### Outports
- *   * __webpage__ GUI elements rendered by web browser.
- *
- * ### Properties
- *   * __URL__ Link to webpage, online or file path.
- *   * __Reload__ Fetch page again.
- *   * __Property__ Type of property to add.
- *   * __Html id__ Identifier of html element to synchronize. Not allowed to contain dots, spaces
- * etc.
- *   * __Add property__ Create a property of selected type and identifier. Start to synchronize
- * against loaded webpage.
  */
 
 /**
  * \brief Render webpage into the color and picking layers (OpenGL).
  */
-#include <warn/push>
-#include <warn/ignore/dll-interface-base>  // Fine if dependent libs use the same CEF lib binaries
-#include <warn/ignore/extra-semi>  // Due to IMPLEMENT_REFCOUNTING, remove when upgrading CEF
-class IVW_MODULE_WEBBROWSER_API WebBrowserProcessor : public Processor, public CefLoadHandler {
+class IVW_MODULE_WEBBROWSER_API WebBrowserProcessor : public Processor, public ProgressBarOwner {
 public:
-    WebBrowserProcessor(InviwoApplication* app);
-    virtual ~WebBrowserProcessor();
+    explicit WebBrowserProcessor(InviwoApplication* app);
 
     virtual void process() override;
 
+    virtual void deserialize(Deserializer& d) override;
+
     virtual const ProcessorInfo getProcessorInfo() const override;
     static const ProcessorInfo processorInfo_;
 
-    void deserialize(Deserializer& d) override;
+private:
+    enum class SourceType { LocalFile, WebAddress };
 
-    // Detect when page has loaded
-    virtual void OnLoadingStateChange(CefRefPtr<CefBrowser> browser, bool isLoading, bool canGoBack,
-                                      bool canGoForward) override;
+    void updateSource();
 
     ImageInport background_;
     ImageOutport outport_;
 
+    OptionProperty<SourceType> sourceType_;
     FileProperty fileName_;
     BoolProperty autoReloadFile_;
-    StringProperty url_;     ///< Web page to show
-    ButtonProperty reload_;  ///< Force reload url
-
+    StringProperty url_;
+    ButtonProperty reload_;
     DoubleProperty zoom_;
-
     ButtonProperty runJS_;
     StringProperty js_;
 
-protected:
-    std::string getSource();
-
-    enum class SourceType { LocalFile, WebAddress };
-
-    OptionProperty<SourceType> sourceType_;
-    BoolProperty isLoading_;
-
-    CEFInteractionHandler cefInteractionHandler_;
-    PickingMapper picking_;
-    CefImageConverter cefToInviwoImageConverter_;
-    // create browser-window
-    CefRefPtr<RenderHandlerGL> renderHandler_;
-    CefRefPtr<CefBrowser> browser_;
-
     SingleFileObserver fileObserver_;
-
-    IMPLEMENT_REFCOUNTING(WebBrowserProcessor);
+    WebBrowserBase browser_;
```

Comment:
How does the ref counting of WebBrowserBase work? I don't know how IMPLEMENT_REFCOUNTING work, But I would expect that you would have to create this object on the heap? with some special pointer to hold it?

Reply:
The macro extends to
```c++
///
/// Macro that provides a reference counting implementation for classes
/// extending CefBase.
///
#define IMPLEMENT_REFCOUNTING(ClassName)          \
 public:                                          \
  void AddRef() const override {                  \
    ref_count_.AddRef();                          \
  }                                               \
  bool Release() const override {                 \
    if (ref_count_.Release()) {                   \
      delete static_cast<const ClassName*>(this); \
      return true;                                \
    }                                             \
    return false;                                 \
  }                                               \
  bool HasOneRef() const override {               \
    return ref_count_.HasOneRef();                \
  }                                               \
  bool HasAtLeastOneRef() const override {        \
    return ref_count_.HasAtLeastOneRef();         \
  }                                               \
                                                  \
 private:                                         \
  CefRefCount ref_count_
```

Reply:
how does this interact with the destructor? seems shady as hell to have this as a member by value...
note the `delete static_cast<const ClassName*>(this);`call

Reply:
https://bitbucket.org/chromiumembedded/cef/wiki/GeneralUsage#markdown-header-reference-counting
need a CefRefPtr I think

Reply:
see https://bitbucket.org/chromiumembedded/cef/wiki/GeneralUsage#markdown-header-reference-counting
use CefRefPtr

Reply:
Nice catch. Seems it was always "broken".

Reply:
yes..